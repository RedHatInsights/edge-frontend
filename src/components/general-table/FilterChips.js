import React, { useEffect } from 'react';
import { ChipGroup, Chip, Button } from '@patternfly/react-core';
import { groupBy } from 'lodash';
import PropTypes from 'prop-types';

const FilterChips = ({
  filterValues,
  setFilterValues,
  chipsArray,
  setChipsArray,
  setPage,
}) => {
  const buildChipsArray = () => {
    filterValues.forEach((filter) => {
      if (filter.type === 'checkbox') {
        filter.value.forEach((filterOption) => {
          if (
            filterOption.isChecked === true &&
            !chipsArray.find((chip) => chip.label === filterOption.option)
          ) {
            setChipsArray((prevState) => [
              ...prevState,
              {
                label: filterOption.option,
                value: filterOption.value || filterOption.option,
                apiName: filterOption.optionApiName,
                key: filter.label,
              },
            ]);
          } else if (
            filterOption.isChecked === false &&
            chipsArray.find((chip) => chip.label === filterOption.option)
          ) {
            setChipsArray((prevState) => {
              const index = prevState.findIndex(
                (state) => state.label === filterOption.option
              );
              return [
                ...prevState.slice(0, index),
                ...prevState.slice(index + 1, prevState.length),
              ];
            });
          }
        });
      }
      if (filter.type === 'text' && filter.value.length > 0) {
        const addTextFilter = { label: filter.value, key: filter.label };
        if (chipsArray.find((chip) => chip.key === filter.label)) {
          setChipsArray((prevState) =>
            prevState.map((f) => {
              return f.key === filter.label ? addTextFilter : f;
            })
          );
        } else {
          setChipsArray((prevState) => [...prevState, addTextFilter]);
        }
      } else if (
        filter.type === 'text' &&
        chipsArray.find((chip) => chip.key === filter.label)
      ) {
        setChipsArray((prevState) => {
          const index = prevState.findIndex(
            (state) => state.key === filter.label
          );
          return [
            ...prevState.slice(0, index),
            ...prevState.slice(index + 1, prevState.length),
          ];
        });
      }
    });
  };

  useEffect(() => {
    buildChipsArray();
    setPage(1);
  }, [filterValues]);

  const handleResetFilters = () => {
    setFilterValues((prevState) => {
      const removedValues = prevState.map((filter) => {
        if (filter.type === 'text') {
          return { ...filter, value: '' };
        }
        if (filter.type === 'checkbox') {
          const setFalse = filter.value.map((checkbox) => {
            checkbox.isChecked = false;
            return { ...checkbox, isChecked: false };
          });
          return { ...filter, value: setFalse };
        }
      });
      return removedValues;
    });
    setChipsArray([]);
  };

  const handleDeleteFilter = (filter) => {
    const filterLabelIndex = filterValues.findIndex(
      (value) => value.label === filter.key
    );
    setFilterValues((prevState) => {
      const changedValue = prevState[filterLabelIndex];
      if (changedValue.type === 'text') {
        return [
          ...prevState.slice(0, filterLabelIndex),
          { ...prevState[filterLabelIndex], value: '' },
          ...prevState.slice(filterLabelIndex + 1, prevState.length),
        ];
      }
      if (changedValue.type === 'checkbox') {
        const changeFalse = changedValue.value.map((option) =>
          option.option === filter.label
            ? { ...option, isChecked: false }
            : option
        );
        setChipsArray((prevState) => {
          const removeIndex = prevState.findIndex((state) => state === filter);
          return [
            ...prevState.slice(0, removeIndex),
            ...prevState.slice(removeIndex + 1, prevState.length),
          ];
        });
        return [
          ...prevState.slice(0, filterLabelIndex),
          { ...prevState[filterLabelIndex], value: changeFalse },
          ...prevState.slice(filterLabelIndex + 1, prevState.length),
        ];
      }
      return prevState;
    });
  };

  return (
    <>
      {chipsArray.length > 0
        ? Object.entries(groupBy(chipsArray, 'key')).map(([key, value]) => (
            <ChipGroup
              data-testid="filter-chip"
              className="pf-u-mr-xs"
              categoryName={key}
              key={key}
            >
              {value.map((filter) => (
                <Chip onClick={() => handleDeleteFilter(filter)} key={key}>
                  {filter.label}
                </Chip>
              ))}
            </ChipGroup>
          ))
        : null}
      {chipsArray.length > 0 ? (
        <Button variant="link" onClick={handleResetFilters}>
          Reset filters
        </Button>
      ) : null}
    </>
  );
};

FilterChips.propTypes = {
  filterValues: PropTypes.array,
  setFilterValues: PropTypes.func,
  chipsArray: PropTypes.array,
  setChipsArray: PropTypes.func,
  setPage: PropTypes.func,
};

export default FilterChips;
