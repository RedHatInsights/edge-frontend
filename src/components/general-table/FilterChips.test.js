import React from 'react';
import FilterChips from './FilterChips';
import { render } from '@testing-library/react';

describe('Filter chips', () => {
  it('should render correctly', async () => {
    const filterValues = [
      { type: 'text', label: 'test-key', value: 'test-label' },
    ];
    const chipsArray = [{ key: 'test-key', label: 'test-label' }];
    const setFilterValues = jest.fn();
    const setChipsArray = jest.fn();
    const setPage = jest.fn();

    const { container, findByTestId, findByText } = render(
      <FilterChips
        filterValues={filterValues}
        setFilterValues={setFilterValues}
        chipsArray={chipsArray}
        setChipsArray={setChipsArray}
        setPage={setPage}
      />
    );

    const FilterChipsElement = await findByTestId('filter-chip');
    const FilterChipsGroupLabel = await findByText('test-key');
    const FilterChipLabel = await findByText('test-label');
    const FilterChipsClearFilters = await findByText('Reset filters');

    expect(FilterChipsElement).toBeDefined();
    expect(FilterChipsGroupLabel).toBeDefined();
    expect(FilterChipLabel).toBeDefined();
    expect(FilterChipsClearFilters).toBeDefined();
    expect(setFilterValues).not.toHaveBeenCalled();
    expect(setChipsArray).toHaveBeenCalled();

    expect(container.querySelector('div')).toMatchSnapshot();
  });
});
