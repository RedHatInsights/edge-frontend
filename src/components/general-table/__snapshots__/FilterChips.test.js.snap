// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter chips should render correctly 1`] = `
<div
  aria-labelledby="pf-random-id-0"
  class="pf-v5-c-chip-group pf-u-mr-xs pf-m-category"
  data-ouia-component-type="PF5/ChipGroup"
  data-ouia-safe="true"
  role="group"
>
  <div
    class="pf-v5-c-chip-group__main"
  >
    <span
      class="pf-v5-c-chip-group__label"
      id="pf-random-id-0"
    >
      test-key
    </span>
    <ul
      aria-labelledby="pf-random-id-0"
      class="pf-v5-c-chip-group__list"
      data-testid="filter-chip"
      role="list"
    >
      <li
        class="pf-v5-c-chip-group__list-item"
      >
        <div
          class="pf-v5-c-chip"
          data-ouia-component-id="OUIA-Generated-Chip-1"
          data-ouia-component-type="PF5/Chip"
          data-ouia-safe="true"
        >
          <span
            class="pf-v5-c-chip__content"
          >
            <span
              class="pf-v5-c-chip__text"
              id="pf-random-id-1"
            >
              test-label
            </span>
          </span>
          <span
            class="pf-v5-c-chip__actions"
          >
            <button
              aria-disabled="false"
              aria-label="close"
              aria-labelledby="remove_pf-random-id-1 pf-random-id-1"
              class="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="close"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              id="remove_pf-random-id-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 352 512"
                width="1em"
              >
                <path
                  d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                />
              </svg>
            </button>
          </span>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
